### PR TITLE
Use IConfiguration to setup endpoint

### LINKED ISSUE #95: Allow endpoint configuration to use IConfiguration
In my Azure Function I am heavily using the `IConfiguration`. More specifically, my connection string to Azure Service Bus is accessed via App Config Service, which in turns gets the value from Key Vault. Ultimately, my actual connection string is available in the `IConfiguration`.  However, based on what I'm seeing it appears that NSB must load the configuration string from an environment variable. See [here](https://github.com/Particular/NServiceBus.AzureFunctions.ServiceBus/blob/b0b734fefa825983bb5d384100bcd71a24a6bb34/src/NServiceBus.AzureFunctions.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs#L32). Ideally, I'd like to pass you a settings 'key' and the value is fetched from an `IConfiguration`.

cc @SeanFeldman

### PR DESCRIPTION
**Pull request by HEskandari**
Closes #95 

More broadly uses IConfiguration to set up the endpoint.

```
class Startup : FunctionsStartup
{
    public override void Configure(IFunctionsHostBuilder builder)
    {
        builder.UseNServiceBus();
    }
}
```

Will need to apply on top of https://github.com/Particular/NServiceBus.AzureFunctions.InProcess.ServiceBus/pull/235 when it is merged.

### CODE REVIEW

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs [hunk 1]
```diff
@@ -18,10 +19,26 @@ static ServiceBusTriggeredEndpointConfiguration()
...
+        public ServiceBusTriggeredEndpointConfiguration(string endpointName, string connectionString = null)
+            : this(endpointName, connectionString, null)
+        {
+        }
+
         /// <summary>
         /// Creates a serverless NServiceBus endpoint running within an Azure Service Bus trigger.
         /// </summary>
-        public ServiceBusTriggeredEndpointConfiguration(string endpointName, string connectionStringName = null)
+        internal ServiceBusTriggeredEndpointConfiguration(string endpointName, string connectionString = null, IConfiguration configuration = null)
```
Comment (SeanFeldman) [inline src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:41]: Note that this will be a breaking change and require a new major since connection string != connection string name.

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs [hunk 2]
```diff
@@ -80,7 +115,7 @@ public static ServiceBusTriggeredEndpointConfiguration FromAttributes()
             var configuration = TriggerDiscoverer.TryGet<ServiceBusTriggerAttribute>();
             if (configuration != null)
             {
-                return new ServiceBusTriggeredEndpointConfiguration(configuration.QueueName, configuration.Connection);
+                return new ServiceBusTriggeredEndpointConfiguration(configuration.QueueName, configuration.Connection); //TODO: How to fix this?
```
Comment (SeanFeldman) [inline src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:118]: You can't. The attribute doesn't provide runtime values, only static, which a connection string **name** is.

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs (hunk 1 above)
Comment (HEskandari) [reply src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:41]: I can undo that to keep the old version.

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs (hunk 2 above)
Comment (HEskandari) [reply src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs]: Should be fixed in the latest commit. I've reverted back to using it as a connection string name.

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs [hunk 3]
```diff
@@ -33,22 +50,21 @@ public ServiceBusTriggeredEndpointConfiguration(string endpointName, string conn
...
             if (!string.IsNullOrWhiteSpace(licenseText))
             {
                 EndpointConfiguration.License(licenseText);
             }
 
             Transport = UseTransport<AzureServiceBusTransport>();
 
-            var connectionString =
-                Environment.GetEnvironmentVariable(connectionStringName ?? DefaultServiceBusConnectionName);
+            var connectionString = GetConfiguredValueOrFallback(configuration, connectionStringName ?? DefaultServiceBusConnectionName);
```
Comment (SeanFeldman) [inline src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:75]: I think this is the scenario where you want to be very explicit and indicate that this configuration is not optional, leading to `bool optional` not having a default value of `false`.

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs [hunk 4]
```diff
@@ -58,6 +74,25 @@ public ServiceBusTriggeredEndpointConfiguration(string endpointName, string conn
             EndpointConfiguration.UseSerialization<NewtonsoftSerializer>();
         }
 
+        private string GetConfiguredValueOrFallback(IConfiguration configuration, string key, bool optional = false)
+        {
+            if (configuration != null)
+            {
+                var configuredValue = configuration.GetValue<string>(key);
+                if (configuredValue != null)
```
Comment (SeanFeldman) [inline src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:90]: null or empty stirng?

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs [hunk 5]
```diff
@@ -58,6 +74,25 @@ public ServiceBusTriggeredEndpointConfiguration(string endpointName, string conn
...
+                {
+                    return configuredValue;
+                }
+            }
+
+            var environmentVariable = Environment.GetEnvironmentVariable(key);
+            if (environmentVariable == null && !optional)
```
Comment (SeanFeldman) [inline src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:97]: null or empty stirng?

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs [hunk 6]
```diff
@@ -58,6 +74,25 @@ public ServiceBusTriggeredEndpointConfiguration(string endpointName, string conn
...
+            {
+                throw new InvalidOperationException($"Configuration or environment value for '{key}' was not set.");
```
Comment (SeanFeldman) [inline src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:99]: was not set or was empty

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs (hunk 4 above)
Comment (HEskandari) [reply src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:90]: I thought about that. I think it will work for now as for all the specific purposes we have non-null or empty string is expected.

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs (hunk 3 above)
Comment (HEskandari) [reply src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:75]: @SeanFeldman not sure I follow, do you mean:

- To change the optional default to false?
- Or pass optional = true for the connection string and throw a more meaningful / explicit exception?

Comment (SeanFeldman) [reply src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs]: I mean not to make `optional` an optional parameter. Rather, be explicit and always pass true or false. It doesn't save much but adds clarity to see what's optional and what's not. Further, I'd also add the parameter name to the call for readability.
E.g.:
```c#
var connectionString = GetConfiguredValueOrFallback(configuration, 
  connectionStringName ?? DefaultServiceBusConnectionName, optional:true);
```

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs (hunk 6 above)
Comment (HEskandari) [reply src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs]: done

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs (hunk 5 above)
Comment (HEskandari) [reply src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs]: done

**Comment by mikeminutillo** [comment]
Is it worth creating an overload of `UseNserviceBus()` that passes `IConfiguration` in to the ServiceBusTriggeredEndpointConfiguration factory method to prevent the closure on IConfiguration and to make it simpler for customers to access?

Additionally, if the configuration can include the endpoint name, would it be worth creating an overload that accepts no parameters and simply loads everything from configuration?

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs [hunk 7]
```diff
@@ -18,10 +19,42 @@ static ServiceBusTriggeredEndpointConfiguration()
             LogManager.UseFactory(FunctionsLoggerFactory.Instance);
         }
 
+        /// <summary>
+        /// Creates a serverless NServiceBus endpoint running within an Azure Service Bus trigger
```
Comment (mikeminutillo) [inline src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:23]: These don't _have_ to exist within an Azure Service Bus trigger. Our own documentation includes [an example of running inside an HTTP trigger](https://docs.particular.net/nservicebus/hosting/azure-functions/service-bus#basic-usage-dispatching-outside-a-message-handler).

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs [hunk 8]
```diff
@@ -33,30 +66,30 @@ public ServiceBusTriggeredEndpointConfiguration(string endpointName, string conn
...
             Transport = UseTransport<AzureServiceBusTransport>();
 
-            var connectionString =
-                Environment.GetEnvironmentVariable(connectionStringName ?? DefaultServiceBusConnectionName);
+            var connectionString = GetConfiguredValueOrFallback(configuration, connectionStringName ?? DefaultServiceBusConnectionName, optional: true);
             if (!string.IsNullOrWhiteSpace(connectionString))
             {
                 Transport.ConnectionString(connectionString);
             }
-            else if (!string.IsNullOrWhiteSpace(connectionStringName))
```
Comment (mikeminutillo) [inline src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:56]: I think this `if` is needed. If we fall back to the `DefaultServiceBusConnectionName` but it's not present, then that's OK. That's exactly what #235 is meant to address. We only throw an exception if you told us to use a specific environment variable but then didn't put anything in it.

File: src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt [hunk 9]
```diff
@@ -53,6 +54,9 @@ namespace NServiceBus
     }
     public class ServiceBusTriggeredEndpointConfiguration
     {
+        public ServiceBusTriggeredEndpointConfiguration(Microsoft.Extensions.Configuration.IConfiguration configuration) { }
+        public ServiceBusTriggeredEndpointConfiguration(string endpointName) { }
+        public ServiceBusTriggeredEndpointConfiguration(string endpointName, Microsoft.Extensions.Configuration.IConfiguration configuration = null) { }
```
Comment (mikeminutillo) [inline src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:59]: Can the compiler tell the difference between this constructor and the one above? I'm surprised we don't get an ambiguous method name compiler error.

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/FunctionsHostBuilderExtensions.cs [hunk 10]
```diff
@@ -3,13 +3,24 @@
...
     /// Provides extension methods to configure a <see cref="FunctionEndpoint"/> using <see cref="IFunctionsHostBuilder"/>.
     /// </summary>
     public static class FunctionsHostBuilderExtensions
     {
+        /// <summary>
+        /// Use the IConfiguration to configures an NServiceBus endpoint that can be injected into a function trigger as a <see cref="FunctionEndpoint"/> via dependency injection.
+        /// </summary>
+        public static void UseNServiceBus(
+            this IFunctionsHostBuilder functionsHostBuilder,
+            IConfiguration configuration)
```
Comment (timbussmann) [inline src/NServiceBus.AzureFunctions.InProcess.ServiceBus/FunctionsHostBuilderExtensions.cs:19]: why do we have to pass the IConfiguration here if we can just grab it from the `IFunctionsHostBuilder` directly?

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs (hunk 7 above)
Comment (HEskandari) [reply src/NServiceBus.AzureFunctions.InProcess.ServiceBus/ServiceBusTriggeredEndpointConfiguration.cs:23]: Updated

File: src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt (hunk 9 above)
Comment (HEskandari) [reply src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:59]: It depends. When calling the constructor, if you do `new ServiceBusTriggeredEndpointConfiguration("test", null)` you do get ambiguous constructor reference, since it can resolve null for both IConfiguration and connectionStringName.

File: src/NServiceBus.AzureFunctions.InProcess.ServiceBus/FunctionsHostBuilderExtensions.cs (hunk 10 above)
Comment (HEskandari) [reply src/NServiceBus.AzureFunctions.InProcess.ServiceBus/FunctionsHostBuilderExtensions.cs:19]: This is fixed in the latest commit.

File: src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt (hunk 9 above)
Comment (mikeminutillo) [reply src/ServiceBus.Tests/ApprovalFiles/APIApprovals.Approve.approved.txt:59]: But if I do `new ServiceBusTriggeredEndpointConfiguration("test")` which constructor gets called? Presumably the one wiithout the nullable parameter. If that's the case, why make this parameter nullable?

**Review by mikeminutillo — APPROVED** [review]
Looks good. I proposed an additional overload here #246

File: src/ServiceBus.Tests/When_no_connection_string_is_provided.cs [hunk 11]
```diff
@@ -12,6 +12,9 @@ public class When_no_connection_string_is_provided
         public void Should_guide_user_towards_success()
         {
             var endpointConfiguration = new EndpointConfiguration("SampleEndpoint");
+            var scanner = endpointConfiguration.AssemblyScanner();
+            scanner.ThrowExceptions = false;
```
Comment (timbussmann) [inline src/ServiceBus.Tests/When_no_connection_string_is_provided.cs:16]: why is this necessary here?

Comment (HEskandari) [reply src/ServiceBus.Tests/When_no_connection_string_is_provided.cs:16]: For some reason, it is necessary everywhere when using Rider. Basically, nunit.engine.dll is picked up by both test runner and assembly scanner and that causes exceptions.

Comment (timbussmann) [reply src/ServiceBus.Tests/When_no_connection_string_is_provided.cs:16]: but it's only configured on this test?
Looking at core, the next major version seems to have the unit.engine dll excluded by default. might be something that we have to configure for the test project if an update of rider doesn't help, but changing this only on a single test sounds off.

Also sounds like something that should be addressed as a separate PR if necessary?

Comment (HEskandari) [reply src/ServiceBus.Tests/When_no_connection_string_is_provided.cs:16]: @timbussmann it was done in a few other places that was broken, via a [different PR](https://github.com/Particular/NServiceBus.AzureFunctions.InProcess.ServiceBus/pull/234).

Comment (timbussmann) [reply src/ServiceBus.Tests/When_no_connection_string_is_provided.cs]: we definitely need to make sure we remove this again in the v8 upgrade because core should exclude those by default. Otherwise we're running into risks of not detecting actual issues with the assembly scanning.

**Review by timbussmann — APPROVED** [review]
